src/core/search/compressed_sorted_set.cc

```diff
@@ -0,0 +1,213 @@
+  auto diff1_span = WriteVarLen(value - bound.prev_value, absl::MakeSpan(buf1));
+  auto diff2_span = WriteVarLen(bound.value - value, absl::MakeSpan(buf2));
+
+  // Extend the location where diff0 is stored with optional zeros before overwriting it
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  size_t required_len = diff1_span.size() + diff2_span.size();
+  DCHECK_LE(bound.diff_span.size(), required_len);  // It can't shrink for sure
+  diffs_.insert(diffs_.begin() + diff_offset, required_len - bound.diff_span.size(), 0u);
+
+  // Now overwrite diff0 and 0s with the two new differences
+  copy(diff1_span.begin(), diff1_span.end(), diffs_.begin() + diff_offset);
+  copy(diff2_span.begin(), diff2_span.end(), diffs_.begin() + diff_offset + diff1_span.size());
+}
+
+// Remove has linear complexity. It tries to find the element V and its neighbors A and B,
+// which are encoded as diff1 = V - A and diff2 = B - V. Adjacently stored diff1 and diff2
+// need to be replaced with diff3 = diff1 + diff2s
+void CompressedSortedSet::Remove(IntType value) {
+  auto bound = LowerBound(value);
+
+  // Nothing was read or the element was not found
+  if (bound.diff_span.empty() || bound.value != value)
+    return;
+
+  // We're removing below unconditionally
+  size_--;
+
+  // Calculate offset where values diff is stored and determine diffs tail
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  auto diffs_tail = absl::MakeSpan(diffs_).subspan(diff_offset + bound.diff_span.size());
+
+  // If it's stored at the end, simply truncate it away
+  if (diffs_tail.empty()) {
+    diffs_.resize(diffs_.size() - bound.diff_span.size());
+    return;
+  }
+
+  // Now the list certainly contains a succeeding element B > V and possibly A < V (or 0)
+  // Read diff2 and calculate diff3 = diff1 + diff2
+  auto [diff2, diff2_read] = ReadVarLen(diffs_tail);
+  IntType diff3 = (bound.value - bound.prev_value) + diff2;
+
+  // Encode diff3
+  VarintBuffer buf;
+  auto diff3_buf = WriteVarLen(diff3, absl::MakeSpan(buf));
+
+  // Shrink vector before overwriting
+  DCHECK_LE(diff3_buf.size(), diff2_read + bound.diff_span.size());
+  size_t to_remove = diff2_read + bound.diff_span.size() - diff3_buf.size();
+  diffs_.erase(diffs_.begin() + diff_offset, diffs_.begin() + diff_offset + to_remove);
+
+  // Overwrite diff1/diff2 with new diff3
+  copy(diff3_buf.begin(), diff3_buf.end(), diffs_.begin() + diff_offset);
+}
+
+size_t CompressedSortedSet::Size() const {
+  return size_;
+}
+
+size_t CompressedSortedSet::ByteSize() const {
+  return diffs_.size();
+}
+
+// The leftmost three bits of the first byte store the number of additional bytes. All following
+// bits store the number itself.
+absl::Span<uint8_t> CompressedSortedSet::WriteVarLen(IntType value, absl::Span<uint8_t> buf) {
+  // TODO: fix flit encoding of large numbers
+  size_t written = base::flit::EncodeT(static_cast<uint64_t>(value), buf.data());
+  return buf.first(written);
```

Comment:
There is a bug with uint32_t values close to their max, so I cast them for now, needs to be fixed in helio